packages/test/test-end-to-end-tests/src/test/sharedTreeEndToEndTests.spec.ts

```diff
@@ -0,0 +1,461 @@
+			await provider.ensureSynchronized();
+		});
+
+		it("can sync initial tree data", async () => {
+			// Validate that the second client received the initial tree data.
+			assert.deepStrictEqual(
+				treeViewClient2.root,
+				initialPoint,
+				"Initial tree data not synchronized",
+			);
+		});
+
+		it("can sync updates to tree data", async () => {
+			// Make changes to the x in first client and y in second client.
+			treeViewClient1.root.x = 3;
+			treeViewClient2.root.y = 4;
+			await provider.ensureSynchronized();
+
+			// Validate that the changes are synchronized.
+			assert.strictEqual(
+				treeViewClient2.root.x,
+				3,
+				"Changes via first client not synchronized",
+			);
+			assert.strictEqual(
+				treeViewClient1.root.y,
+				4,
+				"Changes via second client not synchronized",
+			);
+		});
+	});
+
+	// A slightly complex schema for a shape object that contains an array of points.
+	class Shape extends sf.object("Shape", {
+		sides: sf.number,
+		points: sf.array(Point),
+	}) {}
+
+	function addSide(treeView: TreeView<typeof Shape>, point: Point) {
+		treeView.root.sides++;
+		treeView.root.points.insertAtEnd(point);
+	}
+
+	function removeSide(treeView: TreeView<typeof Shape>) {
+		// treeView.root.sides--;
+		treeView.root.points.removeAt(treeView.root.sides - 1);
+	}
+
+	describe("Tree with an object and array", () => {
+		const treeViewConfig = new TreeViewConfiguration({ schema: Shape });
+		let initialPoint1: Point;
+		let initialPoint2: Point;
+		let treeViewClient1: TreeView<typeof Shape>;
+		let treeViewClient2: TreeView<typeof Shape>;
+
+		beforeEach(async () => {
+			treeViewClient1 = tree1.viewWith(treeViewConfig);
+			treeViewClient2 = tree2.viewWith(treeViewConfig);
+			initialPoint1 = new Point({ x: 1, y: 1 });
+			initialPoint2 = new Point({ x: 2, y: 2 });
+			treeViewClient1.initialize({
+				sides: 2,
+				points: [initialPoint1, initialPoint2],
+			});
+			await provider.ensureSynchronized();
+		});
+
+		it("can sync initial tree data", async () => {
+			// Validate that the second client received the initial tree data.
+			assert.strictEqual(
+				treeViewClient2.root.sides,
+				2,
+				"Initial shape sides not synchronized",
+			);
+			assert.deepStrictEqual(
+				treeViewClient2.root.points.at(0),
+				initialPoint1,
+				"Initial point 1 not synchronized",
+			);
+			assert.deepStrictEqual(
+				treeViewClient2.root.points.at(1),
+				initialPoint2,
+				"Initial point 2 not synchronized",
+			);
+		});
+
+		it("can sync new entries in the array", async () => {
+			// Add a new point in the first client.
+			const newPointClient1 = new Point({ x: 3, y: 3 });
+			addSide(treeViewClient1, newPointClient1);
+			await provider.ensureSynchronized();
+
+			// Validate that the second client received the new point.
+			assert.strictEqual(
+				treeViewClient2.root.points.length,
+				treeViewClient2.root.sides,
+				"Initial shape points not synchronized",
+			);
```

Comment:
Wont this pass even if the edit was not applied?

Its generally hard to tell if the test has these kinds of bugs by just looking at it since the edit is hidden behind addSide.

Just taking a tree with a number in it that starts at 1 and setting it to 2 would make for a way easier to understand and validate test.

Personally, though I think if this test contained the explicit initialization of the tree to one state (so you could see the state inline in the test), then made an edit, then asserted the content of the second tree matched an expected result (I recommend using exportConciseTree and comparing that to a JS literal object with assert.depEqual) it would be much easier to follow and tell if its robust.